routing/index_graph.cpp

```diff
@@ -70,17 +127,29 @@ void IndexGraph::GetNeighboringEdges(Segment const & from, RoadPoint const & rp,
 void IndexGraph::GetNeighboringEdge(Segment const & from, Segment const & to, bool isOutgoing,
                                     vector<SegmentEdge> & edges)
 {
+  // Blocking U-turns on internal feature points.
   RoadPoint const rp = from.GetRoadPoint(isOutgoing);
-  if (from.GetFeatureId() == to.GetFeatureId() && from.GetSegmentIdx() == to.GetSegmentIdx()
-      && from.IsForward() != to.IsForward()
-      && m_roadIndex.GetJointId(rp) == Joint::kInvalidId
-      && rp.GetPointId() != 0
-      && rp.GetPointId() + 1 != m_geometry.GetRoad(from.GetFeatureId()).GetPointsCount())
+  if (IsUTurn(from, to) && m_roadIndex.GetJointId(rp) == Joint::kInvalidId &&
+      rp.GetPointId() != 0 &&
+      rp.GetPointId() + 1 != m_geometry.GetRoad(from.GetFeatureId()).GetPointsCount())
```

Comment:
Actually I'm against such optimizations, this piece of code is quite arguable. From my point of view it's much better to find other bottlenecks.